File: 3.2/conditionalmultiply.swift
import Foundation

func domuls(_ block: String) -> Int {
  var total = 0

  // what is happening here why dont regular regexes work for me i am going insane
  let blockRange = NSRange(
    block.startIndex..<block.endIndex,
    in: block
  )
  let regex = try! NSRegularExpression(pattern: "mul\\((?<l>[0-9]{1,3}),(?<r>[0-9]{1,3})\\)")
  for op in regex.matches(in: block, range: blockRange) {
    var captures: [Int] = []
    for name in ["l", "r"] {
      let matchRange = op.range(withName: name)
      
      if let blockRange = Range(matchRange, in: block) {
        let capture = Int(block[blockRange])!
        captures.append(capture)
      }
    }
    total += captures[0] * captures[1]
  }

  return total
}

let memory = try! String(contentsOfFile: "input.txt", encoding: String.Encoding.utf8)
var total = 0

for ado in memory.split(separator: "do()") {
  let block = String(ado.split(separator: "don't()", maxSplits: 1)[0])
  total += domuls(block)
}

print(total)

File: 3.2/conditionalmultiply.rs
use std::fs::read_to_string;
// need a cargo add to run this one
use regex::Regex;

fn domuls(memory: String) -> i32 {
    let mut total = 0;
    let r = Regex::new(r"mul\(([0-9]{1,3}),([0-9]{1,3})\)").unwrap();

    for (_, [l, r]) in r.captures_iter(&memory).map(|m| m.extract()) {
        total += l.parse::<i32>().unwrap() * r.parse::<i32>().unwrap();
    }

    return total;
}

fn main() {
    let mut memory: String = "".to_owned();
    for line in read_to_string("../input.txt").unwrap().lines() {
        memory.push_str(line);
    }

    let mut ops = Vec::new();
    for ado in memory.split("do()") {
        let dont = ado.split_once("don't()");
         match dont {
            Some(dont) => {
                ops.push(dont.0.to_string());
            },
            None => {
                ops.push(ado.to_string());
            }
         }
    }

    let mut total = 0;
    for op in ops {
        total += domuls(op);
    }

    println!("{:?}", total);
}